presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/GroupArrayAggregationState.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.arrayagg;
+
+import com.facebook.presto.array.IntBigArray;
+import com.facebook.presto.array.LongBigArray;
+import com.facebook.presto.array.ShortBigArray;
+import com.facebook.presto.operator.aggregation.state.AbstractGroupedAccumulatorState;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+import it.unimi.dsi.fastutil.longs.LongArrayList;
+import it.unimi.dsi.fastutil.longs.LongList;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * state object that uses a single BlockBuilder for all groups.
+ */
+public class GroupArrayAggregationState
+        extends AbstractGroupedAccumulatorState
+        implements ArrayAggregationState
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(GroupArrayAggregationState.class).instanceSize();
+    private static final short NULL = -1;
+
+    private final Type type;
+
+    private final ShortBigArray headBlockIndex;
+    private final IntBigArray headPosition;
+
+    private final ShortBigArray nextBlockIndex;
+    private final IntBigArray nextPosition;
+
+    private final LongBigArray tailAbsoluteAddress;
+
+    private final List<BlockBuilder> values;
+    private final LongList sumPositions;
+    private final PageBuilder pageBuilder;
+
+    private long valueBlocksRetainedSizeInBytes;
+    private long totalPositions;
+    private long capacity;
+
+    public GroupArrayAggregationState(Type type)
+    {
+        this.type = type;
+        this.headBlockIndex = new ShortBigArray(NULL);
+        this.headPosition = new IntBigArray(NULL);
+        this.nextBlockIndex = new ShortBigArray(NULL);
+        this.nextPosition = new IntBigArray(NULL);
+        this.tailAbsoluteAddress = new LongBigArray(NULL);
+
+        this.pageBuilder = new PageBuilder(ImmutableList.of(type));
```

Comment:
i'd like to understand PageBuilder sometime, but i can peruse the code.

i'm assuming it behave as implicit in the contract here, so we have a list of BlockBuilders we manage when one gets full, we allocate another.

Reply:
I note the BlockBuilder inside PageBuilder will be capped at 64K, so this is not large enough. Need to discuss with @haozhun  and @dain  on this :).